tests/test_io.py

```diff
@@ -8,86 +8,305 @@
-                    destination,
-                    keep_archive=False,
-                )
-
-        # File should still be there if extraction failed
-        assert os.path.exists(zip_file)
+@pytest.fixture(scope='function', autouse=False)
+def tree(tmpdir, request):
+    r"""Create file tree."""
+
+    files = request.param
+    paths = []
+
+    for path in files:
+        path = path.replace('/', os.path.sep)
+        if path.endswith(os.path.sep):
+            path = audeer.path(tmpdir, path)
+            path = audeer.mkdir(path)
+            path = path + os.path.sep
+            paths.append(path)
+        else:
+            path = audeer.path(tmpdir, path)
+            audeer.mkdir(os.path.dirname(path))
+            path = audeer.touch(path)
+            paths.append(path)
+
+    yield paths
+
+
+@pytest.mark.parametrize(
+    'tree, root, files, archive_create, archive_extract, destination, '
+    'expected',
+    [
+        (  # empty
+            [],
+            '.',
+            [],
+            'archive.zip',
+            'archive.zip',
+            '.',
+            [],
+        ),
+        (
+            [],
+            '.',
+            None,
+            'archive.zip',
+            'archive.zip',
+            '.',
+            [],
+        ),
+        (  # single file
+            ['file.txt'],
+            '.',
+            'file.txt',
+            'archive.zip',
+            'archive.zip',
+            '.',
+            ['file.txt'],
+        ),
+        (  # archive folder does not exist
+            ['file.txt'],
+            '.',
+            'file.txt',
+            'sub/archive.zip',
+            'sub/archive.zip',
+            '.',
+            ['file.txt'],
+        ),
+        (  # file in sub folder
+            ['file.txt', 'sub/a/b/file.txt'],
+            '.',
+            ['sub/a/b/file.txt', 'file.txt'],
+            'archive.zip',
+            'archive.zip',
+            '.',
+            ['sub/a/b/file.txt', 'file.txt'],
+        ),
+        (
+            ['file.txt', 'sub/a/b/file.txt'],
+            '.',
+            ['sub/a/b/file.txt'],
+            'archive.zip',
+            'archive.zip',
+            '.',
+            ['sub/a/b/file.txt'],
+        ),
+        (
+            ['file.txt', 'sub/a/b/file.txt'],
+            '.',
+            ['file.txt'],
+            'archive.zip',
+            'archive.zip',
+            '.',
+            ['file.txt'],
+        ),
+        (  # include all files
+            ['file.txt', 'sub/a/b/file.txt', '.hidden'],
+            '.',
+            None,
+            'archive.zip',
+            'archive.zip',
+            '.',
+            ['.hidden', 'file.txt', 'sub/a/b/file.txt'],
+        ),
+        (  # tar.gz
+            ['file.txt', 'sub/a/b/file.txt'],
+            '.',
+            ['sub/a/b/file.txt', 'file.txt'],
+            'archive.tar.gz',
+            'archive.tar.gz',
+            '.',
+            ['sub/a/b/file.txt', 'file.txt'],
```

Comment:
Is the order of the output given by the order of files, or are folders always listed before files?

Reply:
Yes, that is very unfortunate. But it's the order in which the files are provided to the zip files. The proper solution would have been to sort the files before passing them to the zip file cause now the checksum of the archive depends on the order in which files are passed, e.g.:

``` python
with open('a', 'w') as fp:
    fp.write('a')

with open('b', 'w') as fp:
    fp.write('b')

audeer.create_archive('.', ['a', 'b'], 'ab.zip')
print(audbackend.md5('ab.zip'))

audeer.create_archive('.', ['b', 'a'], 'ba.zip')
print(audbackend.md5('ba.zip'))
```
```
2744ad8eda2f408ac4a84a8a765e4e3a
2505d3354e4c2813aa260839c0f99a60
```

At least with `files=None` we now make sure that files are sorted.

If we want we could at least sort the result of `extract_archive()` if we want to align it with our functions that usually always return files in sorted order.

Reply:
Interestingly, you could also call it a feature.

I would propose the following:
* sort files for `files=None`
* in the docstring state under `files` that the order of the provided files will influence the MD5 sum of the archive and that for `files=None` we sort the files

Regarding sorting at the end I'm not completely sure as it will slow down the process and it might not be needed by every user.

Reply:
Maybe:

![image](https://user-images.githubusercontent.com/10383417/235113856-29860f7f-9ca6-407e-82b5-72d741f70b7b.png)

and:

![image](https://user-images.githubusercontent.com/10383417/235114229-05bf2b69-92e6-4a8c-bd0f-c6fb4d611efe.png)

Reply:
Yes, looks good

Reply:
done